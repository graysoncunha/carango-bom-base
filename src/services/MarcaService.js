import api from '../api/request'

const urlMarcas = '/marcas'

const MarcaService = {
  async cadastrar(marca) {
    return await api(urlMarcas, { method: 'POST', body: marca })
  },

  async alterar(marca) {
    return await api(`${urlMarcas}/${marca.id}`, { method: 'PUT', body: marca })
  },

  async consultar(id) {
    return await api(`${urlMarcas}/${id}`)
  },

  async listar() {
    return await api(urlMarcas)
  },

  async excluir(id) {
    return await api(`${urlMarcas}/${id}`, { method: 'DELETE' })
  },
}

export default MarcaService
