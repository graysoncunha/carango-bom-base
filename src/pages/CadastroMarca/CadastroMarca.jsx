import { Button, TextField, makeStyles } from '@material-ui/core'
import React, { useEffect, useState } from 'react'
import { useHistory, useParams } from 'react-router'
import useErros from '../../hooks/useErros'
import Alert from '@material-ui/lab/Alert'
import MarcaService from '../../services/MarcaService'

const useStyles = makeStyles((theme) => ({
  actions: {
    top: '10px',
    marginRight: '10px',
  },
  alert: {
    width: '100%',
    marginBottom: theme.spacing(2),
  },
}))

function CadastroMarca() {
  const [{ status, error }, setStatus] = useState({ status: 'idle', error: null })
  const [marca, setMarca] = useState('')

  const classes = useStyles()
  const history = useHistory()
  const { id } = useParams()

  const validacoes = {
    marca: (dado) => {
      if (dado && dado.length >= 3) {
        return { valido: true }
      } else {
        return { valido: false, texto: 'Marca deve ter ao menos 3 letras.' }
      }
    },
  }

  const [erros, validarCampos, possoEnviar] = useErros(validacoes)

  function cancelar() {
    history.goBack()
  }

  useEffect(() => {
    if (id) {
      MarcaService.consultar(id).then((m) => setMarca(m.nome))
    }
  }, [id])

  async function cadastrar() {
    try {
      setStatus({ status: 'loading' })
      await MarcaService.cadastrar({ nome: marca })
      setMarca('')
      history.goBack()
      setStatus({ status: 'fulfilled' })
    } catch (e) {
      setStatus({ status: 'rejected', error: e.message })
    }
  }

  return (
    <>
      {status === 'rejected' ? (
        <Alert severity="error" className={classes.alert}>
          {error}
        </Alert>
      ) : null}
      <form
        onSubmit={(event) => {
          event.preventDefault()
          if (possoEnviar()) {
            if (id) {
              MarcaService.alterar({ id, nome: marca }).then(() => {
                history.goBack()
              })
            } else {
              cadastrar()
            }
          }
        }}
      >
        <TextField
          value={marca}
          onChange={(evt) => setMarca(evt.target.value)}
          onBlur={validarCampos}
          helperText={erros.marca.texto}
          error={!erros.marca.valido}
          name="marca"
          id="marca"
          label="Marca"
          type="text"
          variant="outlined"
          fullWidth
          required
          margin="normal"
        />

        <Button variant="contained" color="primary" type="submit" disabled={!possoEnviar()} className={classes.actions}>
          {id ? 'Alterar' : 'Cadastrar'}
        </Button>

        <Button variant="contained" color="secondary" onClick={cancelar} className={classes.actions}>
          Cancelar
        </Button>
      </form>
    </>
  )
}

export default CadastroMarca
