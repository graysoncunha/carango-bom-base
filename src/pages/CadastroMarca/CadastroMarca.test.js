import React from 'react'
import { render, screen, fireEvent, waitFor } from '@testing-library/react'
import { BrowserRouter, MemoryRouter, Route } from 'react-router-dom'
import MarcaService from '../../services/MarcaService'
import CadastroMarca from './CadastroMarca'
import { Router } from 'react-router-dom'
import { createMemoryHistory } from 'history'

const marcaMock = [
  {
    id: 1,
    nome: 'BMW',
  },
]

jest.mock('../../services/MarcaService')

const history = createMemoryHistory()

const renderWithRouter = (ui, { route = '/', path = '/' } = {}) => {
  window.history.pushState({}, 'Test page', route)
  return {
    ...render(
      <BrowserRouter>
        <Route path={path} component={ui} />
      </BrowserRouter>
    ),
  }
}

describe('Componente de cadastro de marcas', () => {
  beforeEach(async () => {
    jest.clearAllMocks()
    MarcaService.listar.mockResolvedValue(marcaMock)
  })
  it('deve carregar a página de cadastro de marca', () => {
    const { container } = render(<CadastroMarca />, { wrapper: MemoryRouter })

    expect(container.firstChild).toMatchSnapshot()
  })

  it('deve carregar a página de cadastro de marca', async () => {
    render(
      <Router history={history}>
        <CadastroMarca />
      </Router>
    )

    const botaoCadastrar = await screen.findByText('Cadastrar')
    expect(botaoCadastrar).toBeInTheDocument()
  })

  it('deve carregar a página de alteração de marca', async () => {
    MarcaService.consultar.mockResolvedValue(marcaMock[0])
    renderWithRouter(CadastroMarca, { route: `/alteracao-marca/${marcaMock[0].id}`, path: '/alteracao-marca/:id' })
    const botaoAlterar = await screen.findByText('Alterar')
    expect(botaoAlterar).toBeInTheDocument()
  })

  it('deve clicar no cancelar da página de alteração de marca', async () => {
    MarcaService.consultar.mockResolvedValue(marcaMock[0])
    renderWithRouter(CadastroMarca, { route: `/alteracao-marca/${marcaMock[0].id}`, path: '/alteracao-marca/:id' })
    const botaoCancelar = await screen.findByText('Cancelar')
    expect(botaoCancelar).toBeInTheDocument()
    fireEvent.click(botaoCancelar)
    expect(history.location.pathname).toBe('/')
  })

  it('quando o formulário é enviado, os campos devem ser validados', () => {
    render(<CadastroMarca />, { wrapper: MemoryRouter })

    fireEvent.click(screen.getByText('Cadastrar'))

    waitFor(() => expect(screen.getByText('Marca deve ter ao menos 3 letras.')).toBeInTheDocument())
  })

  it('deve alterar os campos do formulário', async () => {
    MarcaService.consultar.mockResolvedValue(marcaMock[0])
    renderWithRouter(CadastroMarca, { route: `/alteracao-marca/${marcaMock[0].id}`, path: '/alteracao-marca/:id' })

    const inputNome = await screen.findByDisplayValue(marcaMock[0].nome)
    fireEvent.change(inputNome, { target: { value: 'marca teste' } })
    expect(inputNome.value).toBe('marca teste')
  })
})
