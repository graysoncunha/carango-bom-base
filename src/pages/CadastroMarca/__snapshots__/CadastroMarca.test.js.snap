// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Componente de cadastro de marcas deve carregar a página de cadastro de marca 1`] = `
<form>
  <div
    class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
      data-shrink="false"
      for="marca"
      id="marca-label"
    >
      Marca
      <span
        aria-hidden="true"
        class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
      >
         
        *
      </span>
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input"
        id="marca"
        name="marca"
        required=""
        type="text"
        value=""
      />
      <fieldset
        aria-hidden="true"
        class="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="PrivateNotchedOutline-legendLabelled-5"
        >
          <span>
            Marca
             *
          </span>
        </legend>
      </fieldset>
    </div>
    
  </div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-actions-1 MuiButton-containedPrimary"
    tabindex="0"
    type="submit"
  >
    <span
      class="MuiButton-label"
    >
      Cadastrar
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-actions-1 MuiButton-containedSecondary"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      Cancelar
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</form>
`;
